File: src/clang/magma/compile/rule/tree/TypeRule.c
#include "TypeRule.h"
expand Result_Node_CompileError
expand Result_String_CompileError
struct Result_Node_CompileError parse(struct String input}{return rule.parse(input)
                .mapValue(node -> node.retype(type))
                .mapErr(error -> new CompileError( + type + , new StringContext(input), Lists.of(error)));}struct Result_String_CompileError generate(struct Node node}{if (node.is(type)) {
            return rule.generate(node).mapErr(err -> new CompileError( + type + , new NodeContext(node), Lists.of(err)));
        }

        return new Err<>(new CompileError( + type + , new NodeContext(node)));}

File: src/java/magma/TypeRule.java
package magma;

import magma.api.Err;
import magma.api.Result;
import magma.app.node.Node;
import magma.app.rule.Rule;

public record TypeRule(String type, Rule rule) implements Rule {
    @Override
    public Result<Node, CompileError> lex(String input) {
        return this.rule.lex(input)
                .mapValue(result -> result.retype(this.type));
    }

    @Override
    public Result<String, CompileError> generate(Node node) {
        if (node.is(this.type))
            return this.rule.generate(node);
        return new Err<>(new CompileError("Not of type '" + this.type + "'", new NodeContext(node)));
    }
}
